### PR TITLE
Remove *All APIs from dig

### LINKED ISSUE #40: Remove *All funcs
Background #34 

In order to provide a tight API surface, let's remove the `*All` functions, leaving just:

* `Provide`
* `Resolve`
* `Invoke`

This will mean combining the `ProvideAll` and `ResolveAll` into `Provide` and `Resolve`, so that those functions deal with variadic arguments so they can operate on one or many deps.

### PR DESCRIPTION
**Pull request by anuptalwalkar**
resolves #40 
`Provide`, `Resolve` APIs support variadic arguments to reduce API surface area to `Provide`, `Resolve` and `Invoke`

### CODE REVIEW

File: container.go [hunk 1]
```diff
@@ -83,85 +83,72 @@ func (c *Container) Invoke(t interface{}) error {
...
-	switch ctype.Kind() {
-	case reflect.Func:
-		switch ctype.NumOut() {
-		case 0:
-			return errReturnCount
-		case 1:
-			objType := ctype.Out(0)
-			if objType.Kind() != reflect.Ptr && objType.Kind() != reflect.Interface {
-				return errReturnKind
+func (c *Container) Provide(types ...interface{}) error {
```
Comment (alsamylkin) [inline container.go]: This means DIG will not support any tagging with names in Provide in the future, right?

Comment (breerly) [inline container.go]: That's right - this API will become single purpose for the time being.

File: container.go [hunk 2]
```diff
@@ -83,85 +83,72 @@ func (c *Container) Invoke(t interface{}) error {
...
+			case 1:
+				objType := ctype.Out(0)
+				if objType.Kind() != reflect.Ptr && objType.Kind() != reflect.Interface {
+					return errReturnKind
+				}
 			}
+			if err := c.Graph.InsertConstructor(t); err != nil {
+				return err
+			}
+		case reflect.Slice, reflect.Array, reflect.Map, reflect.Ptr:
```
Comment (alsamylkin) [inline container.go]: Array is more a value type not a pointer like Slice/Map/Interface, not sure if it should be here.

Comment (alsamylkin) [inline container.go]: surprised why channels are not there.

**Pull request merged**